src/rustbox.rs

```diff
@@ -272,11 +271,16 @@ pub struct RustBox {
     // Note that running *MUST* be the last field in the destructor, since destructors run in
     // top-down order. Otherwise it will not properly protect the above fields.
     _running: running::RunningGuard,
-    // Termbox is not thread safe. See #39.
-    _phantom: PhantomData<UnsafeCell<()>>,
 
     // Store this so we know which colours to use
     output_mode: OutputMode,
+
+    // Termbox is partially thread-safe. These locks allow Rustbox to
+    // synchronize access to non-thread-safe method groups (enabling listening
+    // for input events while writing output to the terminal).
+    // See https://github.com/nsf/termbox/commit/493dc1395c91174e97658ff15fa2380227faf28f
+    input_lock: Mutex<()>,
```

Comment:
Initially, the intent was to keep a simple separation: `input_lock` for anything where we're reading from the terminal, and `output_lock` for anything where we're writing to the terminal (or termbox's internal buffer). What we're primarily trying to avoid is two threads modifying termbox's internal state, since those writes aren't atomic. I've grouped `set_input_mode` and `set_output_mode` into `output_lock` because they modify state, even though the former sounds like a good fit for `input_lock`.

Slightly out of place, `height` and `width` might seem better off using `input_lock` (as a resize event would be the only thing that would change their values), but in a typical application, where an input listener is holding the input lock, and the renderer is using `width/height`, that either degrades performance (as each call in the renderer waits for a lock), or breaks the model altogether (if the input listener blocks indefinitely on input). I'd be inclined to drop the lock completely in that case, if we changed it at all. :slightly_smiling_face:

Reply:
Thanks for the explanation! Makes sense.

Could you add a brief explanation like this to the field definitions? Future me probably won't think to refer back to this comment, and while I'd like to think I remember the distinction I probably won't :/  Just a comment for each field to say what each one is for would be great. Essentially your first sentence above, just in the code instead of a Github comment.

Looks great though! Thanks for taking the time to contribute!

Reply:
Of course, happy to take care of that tonight!

Likewise, this library has been great; appreciate all of the hard work you've put into it. :beers:

Reply:
Thanks!

And thank you for the encouragement. Lots of folks have helped get it to this point which is great. I'm glad to hear it has been useful! I would love to have more time to move it towards being a pure-rust implementation, but perhaps one day!